### PR TITLE
upx 3.95

### PR DESCRIPTION
**Pull request by javian**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
New release with patch to make it work on MacOS.

### CODE REVIEW

**Comment by javian** [comment]
@BrewTestBot test this please

File: Formula/upx.rb [hunk 1]
```diff
@@ -26,6 +15,13 @@ class Upx < Formula
 
   depends_on "ucl"
 
+  # Patch required due to 3.95 MacOS bug https://github.com/upx/upx/issues/218
+  # and ought to be included in the next release
+  patch do
+    url "https://github.com/upx/upx/commit/0dac6b7be3339ac73051d40ed4d268cd2bb0dc7c.patch?full_index=1"
+    sha256 "71f4048ba4edc4c4530ebdce63193503430d75110f35f0989726b7cfb00867f7"
```
Comment (fxcoudert) [inline Formula/upx.rb]: Should be `957de8bab55bb71156a1ae59fa66c67636acd265a4c6fa43d12e8793bafebb22`

**Comment by fxcoudert** [comment]
Failure on El Capitan only:

```
==> /usr/local/Cellar/upx/3.95/bin/upx -1 ./upx
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2018
UPX 3.95        Markus Oberhumer, Laszlo Molnar & John Reiser   Aug 26th 2018

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
   1933268 ->    671760   34.75%   macho/amd64   upx

Packed 1 file.
==> ./upx -V
00000000000000c5=000000000000000c
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:49:in `write': Broken pipe (Errno::EPIPE)
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:49:in `puts'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:49:in `rescue in block (3 levels) in safe_fork'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:31:in `block (3 levels) in safe_fork'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:30:in `fork'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:30:in `block (2 levels) in safe_fork'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:27:in `open'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:27:in `block in safe_fork'
	from /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/tmpdir.rb:89:in `mktmpdir'
	from /usr/local/Homebrew/Library/Homebrew/utils/fork.rb:26:in `safe_fork'
	from /usr/local/Homebrew/Library/Homebrew/dev-cmd/test.rb:80:in `block in test'
	from /usr/local/Homebrew/Library/Homebrew/dev-cmd/test.rb:29:in `each'
	from /usr/local/Homebrew/Library/Homebrew/dev-cmd/test.rb:29:in `test'
	from /usr/local/Homebrew/Library/Homebrew/brew.rb:89:in `<main>'
Error: upx: failed
undefined method `map' for nil:NilClass
Did you mean?  tap
/usr/local/Homebrew/Library/Homebrew/exceptions.rb:363:in `initialize'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:13:in `new'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:13:in `rewrite_child_error'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:78:in `block (3 levels) in safe_fork'
/usr/local/Homebrew/Library/Homebrew/utils.rb:399:in `ignore_interrupts'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:57:in `block (2 levels) in safe_fork'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:27:in `open'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:27:in `block in safe_fork'
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/tmpdir.rb:89:in `mktmpdir'
/usr/local/Homebrew/Library/Homebrew/utils/fork.rb:26:in `safe_fork'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/test.rb:80:in `block in test'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/test.rb:29:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/test.rb:29:in `test'
/usr/local/Homebrew/Library/Homebrew/brew.rb:89:in `<main>'
```

**Comment by javian** [comment]
Aarrgh... =| I don't have an el cap machine to this this on right now but I'll see what I can dig up over the weekend.

**Comment by scpeters** [comment]
I have an el cap machine at work; I'll give it a try

**Comment by scpeters** [comment]
I can reproduce the failure, but I'm not sure how to debug further:

~~~
$ /usr/local/bin/upx -V && echo success
upx 3.95
UCL data compression library 1.03
zlib data compression library 1.2.5
LZMA SDK version 4.43
Copyright (C) 1996-2018 Markus Franz Xaver Johannes Oberhumer
Copyright (C) 1996-2018 Laszlo Molnar
Copyright (C) 2000-2018 John F. Reiser
Copyright (C) 2002-2018 Jens Medoch
Copyright (C) 1995-2005 Jean-loup Gailly and Mark Adler
Copyright (C) 1999-2006 Igor Pavlov
UPX comes with ABSOLUTELY NO WARRANTY; for details type 'upx -L'.
success
~~~

~~~
$ ./upx -V && echo success
00000000000000c5=000000000000000c
~~~

the compressed binary doesn't work the same. it prints out something garbled and returns an error code

**Comment by scpeters** [comment]
The weird thing is that I copied the compressed `upx` binary that fails on el cap to my high sierra machine and it works on high sierra.

**Comment by scpeters** [comment]
big diff https://github.com/upx/upx/compare/v3.94...v3.95

**Comment by scpeters** [comment]
The el cap binary can be downloaded from the jenkins bottle if you want to confirm my finding that it works on newer OS versions:

* https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/30127/version=el_capitan/artifact/upx--3.95.el_capitan.bottle.tar.gz

**Comment by javian** [comment]
This is an upstream issue and needs to be reported in a similar fashion like I did with what resulted in the patch https://github.com/upx/upx/issues/218.

**Comment by jreiser** [comment]
The first step in debugging is to get a list of the system calls that were made, particularly for the address space layout (mmap, etc.)  See https://stackoverflow.com/questions/31045575/how-to-trace-system-calls-of-a-program-in-mac-os-x and others.  The second step might be to run under lldb, and single-step the beginning of execution.  Skip over the actual de-compression subroutine (called a few times), but pay attention to the "supervision".  Stop at the transfer to the original code; look at the address space layout, dump memory to a file.  Compare with the analogous dump of the non-compressed program at its first instruction.

**Comment by jreiser** [comment]
The output "00000000000000c5=000000000000000c" is an internal debugging message from upx that indicates that syscall 0xc5 (mmap) encountered errno 0xc (ENOMEM).  The code is in src/stub/src/amd64-darwin.macho-fold.S at label sysfail.
Dtruss is USELESS because friggin' Apple decided that nobody would ever need to list the system calls from a "bare" program:  "dtrace: failed to execute ./foo: Could not create symbolicator for task".  This is why Linux is a vastly superior development environment: 'strace' just works; it needs no symbols in the monitored process.
Putting a breakpoint at the 'syscall' instruction, and watching all syscalls made after the bootstrapping phase of decompression, shows that the problematic call is mmap(0x0000fffebf838000, 0x0000000000003000, 0x0000000000000003, 0x0000000000000012, 0x0000000000000004, 0x0000000000039000); for compressed /bin/date from ElCapitan: I found an old system.  There is no memory available at the  first argument address 0x0000fffebf838000.  Now to figure out why that address was chosen, and only on ElCapitan.
Hint: (lldb) process launch -s

**Comment by jreiser** [comment]
El Capitan (MacOS 10.11.x) has a "pre-linked" /usr/lib/dyld.  Should be fixed in  upx 9bb6854e642a2505102b9d3f9ec8535ec8ab6d9c .

**Comment by javian** [comment]
Thanks @jreiser - however it seems that the *-darwin.mach-fold.map files does not exist in the src/stub/tmp directory in the release tarball
```
can't find file to patch at input line 853
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/src/stub/tmp/amd64-darwin.macho-fold.map b/src/stub/tmp/amd64-darwin.macho-fold.map
|index 9c89202dc083ee3650e0d15fb5cf72815bdd4c62..721815637b3a613d756be9b9cde7557e3ca13276 100644
|--- a/src/stub/tmp/amd64-darwin.macho-fold.map
|+++ b/src/stub/tmp/amd64-darwin.macho-fold.map
--------------------------
No file to patch.  Skipping patch.
5 out of 5 hunks ignored
can't find file to patch at input line 984
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/src/stub/tmp/arm.v5a-darwin.macho-fold.map b/src/stub/tmp/arm.v5a-darwin.macho-fold.map
|index 684c9c41083f46ca2518158d502a7d8af8819cb1..58cb4f1a33e097de0cfd565fd821b6fb2beb025e 100644
|--- a/src/stub/tmp/arm.v5a-darwin.macho-fold.map
|+++ b/src/stub/tmp/arm.v5a-darwin.macho-fold.map
--------------------------
No file to patch.  Skipping patch.
5 out of 5 hunks ignored
can't find file to patch at input line 1115
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/src/stub/tmp/arm64-darwin.macho-fold.map b/src/stub/tmp/arm64-darwin.macho-fold.map
|index 4252fcbe68f357112f619237ed8c5f608e0a1c5b..94d4c5081ef2b2d345798ce261c86b463cbddbb4 100644
|--- a/src/stub/tmp/arm64-darwin.macho-fold.map
|+++ b/src/stub/tmp/arm64-darwin.macho-fold.map
--------------------------
No file to patch.  Skipping patch.
5 out of 5 hunks ignored
```

**Comment by javian** [comment]
Or hmm, the files are there - just not sure why they don't patch well yet =/

**Comment by javian** [comment]
From the tarball
```
➜  tmp ls -la *darwin*.map
-rw-r--r--  1 javian  staff  6567 Aug 26 04:58 amd64-darwin.macho-fold.map
-rw-r--r--  1 javian  staff  6758 Aug 26 04:58 arm.v5a-darwin.macho-fold.map
-rw-r--r--  1 javian  staff  8046 Aug 26 04:58 arm64-darwin.macho-fold.map
-rw-r--r--  1 javian  staff  6498 Aug 26 04:58 i386-darwin.macho-fold.map
-rw-r--r--  1 javian  staff  6304 Aug 26 04:58 powerpc-darwin.macho-fold.map
-rw-r--r--  1 javian  staff  9460 Aug 26 04:58 powerpc64le-darwin.macho-fold.map
➜  tmp pwd
/upx-3.95/src/stub/tmp
```

**Comment by DomT4** [comment]
```diff
diff --git a/Formula/upx.rb b/Formula/upx.rb
index dcfc936aa9..37f9e89b26 100644
--- a/Formula/upx.rb
+++ b/Formula/upx.rb
@@ -1,10 +1,34 @@
 class Upx < Formula
   desc "Compress/expand executable files"
   homepage "https://upx.github.io/"
-  url "https://github.com/upx/upx/releases/download/v3.95/upx-3.95-src.tar.xz"
-  sha256 "3b0f55468d285c760fcf5ea865a070b27696393002712054c69ff40d8f7f5592"
   head "https://github.com/upx/upx.git", :branch => :devel
 
+  stable do
+    url "https://github.com/upx/upx/archive/v3.95.tar.gz"
+    sha256 "fdb79c8238360115770e9c13bdaeb48da6fb09c813b0a461c5f9faee176d6fb9"
+
+    # This can be removed, along with the two patches, when switching back to
+    # the release tarball rather than the archive tarball. This should be
+    # done on the next stable release.
+    resource "lzma-sdk" do
+      url "https://github.com/upx/upx-lzma-sdk/archive/v3.95.tar.gz"
+      sha256 "4932ed7b79cf47aa91fe737c068f74553e17033161c7e7e532e4b967f02f1557"
+    end
+
+    # Patch required due to 3.95 MacOS bug https://github.com/upx/upx/issues/218
+    # and ought to be included in the next release
+    patch do
+      url "https://github.com/upx/upx/commit/0dac6b7be3339ac73051d40ed4d268cd2bb0dc7c.patch?full_index=1"
+      sha256 "957de8bab55bb71156a1ae59fa66c67636acd265a4c6fa43d12e8793bafebb22"
+    end
+
+    # https://github.com/Homebrew/homebrew-core/pull/31846#issuecomment-419750313
+    patch do
+      url "https://github.com/upx/upx/commit/9bb6854e642a2505102b9d3f9ec8535ec8ab6d9c.patch?full_index=1"
+      sha256 "f525a574b65e6484f0eb29e2a37d5df58da85b121adec06271b19ed5f4cc49b4"
+    end
+  end
+
   bottle do
     cellar :any_skip_relocation
     sha256 "d45aa693d18de2592fa5e776e042f22f7e13180ac463f130991c3ce815f9d3c1" => :mojave
@@ -15,14 +39,8 @@ class Upx < Formula
 
   depends_on "ucl"
 
-  # Patch required due to 3.95 MacOS bug https://github.com/upx/upx/issues/218
-  # and ought to be included in the next release
-  patch do
-    url "https://github.com/upx/upx/commit/0dac6b7be3339ac73051d40ed4d268cd2bb0dc7c.patch?full_index=1"
-    sha256 "957de8bab55bb71156a1ae59fa66c67636acd265a4c6fa43d12e8793bafebb22"
-  end
-
   def install
+    (buildpath/"src/lzma-sdk").install resource("lzma-sdk") if build.stable?
     system "make", "all"
     bin.install "src/upx.out" => "upx"
     man1.install "doc/upx.1"
```

**Comment by DomT4** [comment]
screenshot 2018-09-10 at 21 36 55

Thank you GitHub for immortalising the fact I forgot to `git reset` my local clone before pushing the latest `master` changes to my fork 😅.

**Review by DomT4 — APPROVED** [review]
Thanks for picking this one up @javian ❤️

**Comment by javian** [comment]
And we're all green now and lol @DomT4 - you're too quick for your own good =)

**Comment by javian** [comment]
@jreiser thanks for taking the time to have a look at this so quickly and we're glad to ship your project !

**Pull request closed without merging**